### ISSUE TITLE
Writing the cache control file in `ResourceCache` is not thread-safe, resulting in fatal errors being printed

### ISSUE THREAD

**Issue by Vampire** [xmlresolver/xmlresolver#155]
We have an ear that is deployed to a WildFly instance.
In this ear there are multiple rest easy applications that are initialized during startup in multiple worker threads.
Sometimes it happens that a fatal error is logged about `control.xml`.

After some debugging and trying, I found that the problem is in this project.
Those rest easy applications cause a `DocumentProvider` of rest easy to be created,
which calls `TransformerFactory.newInstance()`,
which instantiates a Saxon transformer factory as we have saxon as library,
which through some callchain in the end creates a new `ResourceCache`.

With the catalog cache enabled and those calls happening on multiple threads concurrently, with the file not being present at start of execution as all this is running in a fresh Docker instance, occasionally the Fatal error is logged, that the `config.xml` prematurely ended, or that it does not start and end with the same tag or any other parsing error.

The problem is, that the first thread sees the file is missing and starts writing it.
While it is half-way through with writing it, the second thread sees the file exists and tries to parse it.
But as the file is only half-way written, the parsing fails and a "Fatal Error" is logged about it.

The error can easily be reproduced in a standalone main class like this:
```java
System.setProperty("xml.catalog.cacheEnabled", "true"); // needed since xmlresolver 4.5.0 as caching is disabled by default now
var executorService = Executors.newFixedThreadPool(2);
try {
    while (true) {
        Files.deleteIfExists(Path.of(System.getProperty("user.home"), ".xmlresolver.org/cache/control.xml"));
        var latch = new CountDownLatch(2);
        Callable<Configuration> createConfiguration = () -> {
            latch.countDown();
            latch.await();
            return Configuration.newConfiguration();
        };
        var x = executorService.submit(createConfiguration);
        var y = executorService.submit(createConfiguration);
        x.get();
        y.get();
    }
} finally {
    executorService.shutdown();
}
```
I first reproduced it with the versions we currently use in our application, which are Saxon-HE 11.3 and xmlresolver 4.2.0.
There it reproduces occasionally with the above reproducer (as in once every few seconds).
But it happens exactly the same with Saxon-HE 12.3 and xmlresolver 5.2.2.
The only differences are, that you now have to enable the caching as it is disabled by default since xmlresolver 4.5.0 and that it reproduces much more often than with the old versions.

I guess you should either have some filesystem level lock or marker file while writing the file to wait with the parsing, or first assemble the file with a different name and then move it into the final location after it was fully written, so that the chance a parser sees an incomplete file is minimized.

**Comment by ndw** [comment]
I've come to the conclusion that the whole automatic caching thing was a complete failure. I've pulled it out of version 6, which I hope to publish "this month". I suggest that you just turn off caching. Sorry about the thread safety issue, though.

**Comment by Vampire** [comment]
Well, was far as I have seen, it then just uses the defaults, which it otherwise would have tried to write to the file or read from the file and the only itch is the unnecessary "Fatal Error", which will anyway not come at our customers, as none as far as I know uses Docker or similar and thus the file should simply be there already. I just found it on internal test systems that run on Docker. So I decided to just ignore it until we update to 4.5.0+ where caching is disabled by default anyway, thanks. :-)
